analyzers/src/SonarAnalyzer.CSharp/Rules/ReturnEmptyCollectionInsteadOfNull.cs

```diff
@@ -18,68 +18,70 @@
-                var nullOrDefaultLiterals = GetNullOrDefaultExpressions(expressionBody.Expression)
-                    .Select(statement => statement.GetLocation())
-                    .ToList();
+            var nullOrDefaultLiterals = GetNullOrDefaultExpressions(expressionBody.Expression)
+                .Select(statement => statement.GetLocation())
+                .ToList();
 
-                ReportIfAny(nullOrDefaultLiterals);
-            }
-            else if (GetBody(context.Node) is { } body)
-            {
-                var nullOrDefaultLiterals = GetReturnNullOrDefaultExpressions(body)
-                    .Select(returnStatement => returnStatement.GetLocation())
-                    .ToList();
+            ReportIfAny(nullOrDefaultLiterals);
+        }
+        else if (GetBody(context.Node) is { } body)
+        {
+            var nullOrDefaultLiterals = GetReturnNullOrDefaultExpressions(body)
+                .Select(returnStatement => returnStatement.GetLocation())
+                .ToList();
 
-                ReportIfAny(nullOrDefaultLiterals);
-            }
+            ReportIfAny(nullOrDefaultLiterals);
+        }
 
-            void ReportIfAny(List<Location> nullOrDefaultLiterals)
+        void ReportIfAny(List<Location> nullOrDefaultLiterals)
+        {
+            if (nullOrDefaultLiterals.Count > 0 && IsReturningCollection(context))
             {
-                if (nullOrDefaultLiterals.Count > 0 && IsReturningCollection(context))
-                {
-                    context.ReportIssue(Rule.CreateDiagnostic(context.Compilation, nullOrDefaultLiterals[0], additionalLocations: nullOrDefaultLiterals.Skip(1)));
-                }
+                context.ReportIssue(Rule.CreateDiagnostic(context.Compilation, nullOrDefaultLiterals[0], additionalLocations: nullOrDefaultLiterals.Skip(1)));
             }
         }
+    }
 
-        private static bool IsReturningCollection(SonarSyntaxNodeReportingContext context)
-        {
-            var symbol = context.SemanticModel.GetDeclaredSymbol(context.Node);
-            var methodSymbol = (symbol as IPropertySymbol)?.GetMethod ?? symbol as IMethodSymbol;
+    private static bool IsReturningCollection(SonarSyntaxNodeReportingContext context) =>
+        GetReturnMethod(context) is { } method
+        && !method.ReturnType.Is(KnownType.System_String)
+        && !method.ReturnType.DerivesFrom(KnownType.System_Xml_XmlNode)
+        && method.ReturnType.DerivesOrImplementsAny(CollectionTypes)
+        && method.ReturnType.NullableAnnotation() != NullableAnnotation.Annotated;
 
-            return methodSymbol != null
-                && !methodSymbol.ReturnType.Is(KnownType.System_String)
-                && !methodSymbol.ReturnType.DerivesFrom(KnownType.System_Xml_XmlNode)
-                && methodSymbol.ReturnType.DerivesOrImplementsAny(CollectionTypes);
-        }
+    private static IMethodSymbol GetReturnMethod(SonarSyntaxNodeReportingContext context)
```

Comment:
This method can be inlined.

Reply:
I'm not sure if you thought about the same change, but I've changed the method to bet `GetType()`, and I think it makes matters even easier.